pkg/ha/ha_service.go

```diff
@@ -0,0 +1,191 @@
+package ha
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgconn"
+	"github.com/timescale/promscale/pkg/ha/state"
+
+	"golang.org/x/sync/semaphore"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const (
+	backOffDurationOnLeaderChange = 10 * time.Second
+	haSyncerTimeInterval          = 15 * time.Second
+)
+
+type Service struct {
+	state             *sync.Map
+	lockClient        haLockClient
+	leaseTimeout      time.Duration
+	leaseRefresh      time.Duration
+	_leaderChangeLock *semaphore.Weighted
+	_syncTicker       *time.Ticker
+}
+
+func NewHAService(dbClient pgxconn.PgxConn) (*Service, error) {
+	lockClient := newHaLockClient(dbClient)
+	timeout, refresh, err := lockClient.readLeaseSettings(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("could not create HA state: %#v", err)
+	}
+
+	service := &Service{
+		state:             &sync.Map{},
+		lockClient:        lockClient,
+		leaseTimeout:      timeout,
+		leaseRefresh:      refresh,
+		_leaderChangeLock: semaphore.NewWeighted(1),
+		_syncTicker:       time.NewTicker(haSyncerTimeInterval),
+	}
+	go service.haStateSyncer()
+	return service, nil
+}
+
+func (h *Service) haStateSyncer() {
+	for range h._syncTicker.C {
+		h.state.Range(func(c, s interface{}) bool {
+			cluster := fmt.Sprint(c)
+			state := castToState(s)
+			stateView := state.Clone()
+			err := h.syncLockStateFromDB(stateView.LeaseStart, stateView.MaxTimeSeenLeader, cluster, stateView.Leader)
+			if err != nil {
+				log.Error("failed to validate cluster %s state %v", cluster, err)
+			}
+			if stateView.MaxTimeSeen.After(stateView.LeaseUntil) {
```

Comment:
can we factor this boolean expression into a function since it used in 2 places. here and on line 156. Plus we may want to adjust these.

Reply:
If I am getting it right the check here is

we do this check before calling tryChangeLeader `if stateView.MaxTimeSeen.After(stateView.LeaseUntil)` and check 156 is 
`stateView.LeaseUntil.After(stateView.MaxTimeSeen)` where this check is to re-verify before changing the leader. 

DO you mean we should create another generic func for this comparison? Also, can you elaborate on adjusting this?